src/mlpack/core/data/dataset_info.hpp

```diff
@@ -13,37 +13,29 @@
 #include <unordered_map>
 #include <boost/bimap.hpp>
 
+#include "map_policies/increment_policy.hpp"
+
 namespace mlpack {
 namespace data {
-
-/**
- * The Datatype enum specifies the types of data mlpack algorithms can use.  The
- * vast majority of mlpack algorithms can only use numeric data (i.e.
- * float/double/etc.), but some algorithms can use categorical data, specified
- * via this Datatype enum and the DatasetInfo class.
- */
-enum Datatype : bool /* bool is all the precision we need for two types */
-{
-  numeric = 0,
-  categorical = 1
-};
-
 /**
  * Auxiliary information for a dataset, including mappings to/from strings and
- * the datatype of each dimension.  DatasetInfo objects are optionally produced
- * by data::Load(), and store the type of each dimension (Datatype::numeric or
- * Datatype::categorical) as well as mappings from strings to unsigned integers
- * and vice versa.
+ * the datatype of each dimension.  DatasetMapper objects are optionally
+ * produced by data::Load(), and store the type of each dimension
+ * (Datatype::numeric or Datatype::categorical) as well as mappings from strings
+ * to unsigned integers and vice versa.
+ *
+ * @tparam MapPolicy Mapping policy used to specify MapString();
  */
-class DatasetInfo
+template <typename MapPolicy>
+class DatasetMapper
 {
  public:
   /**
-   * Create the DatasetInfo object with the given dimensionality.  Note that the
-   * dimensionality cannot be changed later; you will have to create a new
-   * DatasetInfo object.
+   * Create the DatasetMapper object with the given dimensionality.  Note that
+   * the dimensionality cannot be changed later; you will have to create a new
+   * DatasetMapper object.
    */
-  DatasetInfo(const size_t dimensionality = 0);
+  DatasetMapper(const size_t dimensionality = 0);
```

Comment:
Let the constructor store MapPolicy, so we could save some state and allow us to tell the DatasetMapper we do not want to map non target string(ex : NaN, empty string and so on) to categorical.

```
DatasetMapper(const size_t dimensionality = 0, MapPolicy mapPolicy) : 
  dimensionality(dimensionality),
   mapPolicy(std::move(mapPolicy))
{}
```